utils/gridlink_impl.c.src

```diff
@@ -179,113 +142,163 @@ cellarray_DOUBLE * gridlink_DOUBLE(const int64_t NPART,
         if (ix>nmesh_x-1)  ix--;    /* this shouldn't happen, but . . . */
         if (iy>nmesh_y-1)  iy--;
         if (iz>nmesh_z-1)  iz--;
-        XRETURN(X[i] >= xmin && X[i] <= xmax, NULL,
-               "x[%"PRId64"] = %"REAL_FORMAT" must be within [%"REAL_FORMAT",%"REAL_FORMAT"]\n",
-                i, X[i], xmin, xmax);
-        XRETURN(Y[i] >= ymin && Y[i] <= ymax, NULL,
-               "y[%"PRId64"] = %"REAL_FORMAT" must be within [%"REAL_FORMAT",%"REAL_FORMAT"]\n",
-               i, Y[i], ymin, ymax);
-        XRETURN(Z[i] >= zmin && Z[i] <= zmax, NULL,
-               "z[%"PRId64"] = %"REAL_FORMAT" must be within [%"REAL_FORMAT",%"REAL_FORMAT"]\n",
-               i, Z[i], zmin, zmax);
-
-        XRETURN(ix >= 0 && ix < nmesh_x, NULL, "ix=%d must be within [0,%d)\n", ix, nmesh_x);
-        XRETURN(iy >= 0 && iy < nmesh_y, NULL, "iy=%d must be within [0,%d)\n", iy, nmesh_y);
-        XRETURN(iz >= 0 && iz < nmesh_z, NULL, "iz=%d must be within [0,%d)\n", iz, nmesh_z);
+    
+        out_of_bounds += ix < 0 || ix >= nmesh_x || iy < 0 || iy >= nmesh_y || iz < 0 || iz >= nmesh_z;
 
         const int64_t icell = CONVERT_3D_INDEX_TO_LINEAR(ix, iy, iz, nmesh_x, nmesh_y, nmesh_z);
-        if(options->copy_particles == 0) {
-            all_cell_indices[i] = icell;
+        all_cell_indices[i] = icell;
+        if(!options->copy_particles){
             original_indices[i] = i;
+        }
+    }
+    
+    XRETURN(out_of_bounds == 0, NULL, "%"PRId64" particles are out of bounds. Check periodic wrapping?\n", out_of_bounds);
+    
+    // Now determine the number of particles per cell so each cell knows its global starting index
+    // This information is used for both the copy and in-place versions
+#if defined(_OPENMP)
+    int max_nthreads = omp_get_max_threads();
+#else
+    int max_nthreads = 1;
+#endif
+    int64_t *cell_occupation[max_nthreads];
+
+#if defined(_OPENMP)
+    #pragma omp parallel
+#endif
+    {
+        #if defined(_OPENMP)
+        int tid = omp_get_thread_num();
+        #else
+        int tid = 0;
+        #endif
+        cell_occupation[tid] = my_calloc(totncells, sizeof(int64_t));
+        
+        #if defined(_OPENMP)
+        #pragma omp for schedule(static)
+        #endif
+        for(int64_t i = 0; i < NPART; i++){
+            cell_occupation[tid][all_cell_indices[i]]++;
+        }
+    }
+    
+#if defined(_OPENMP)
+    #pragma omp parallel for schedule(static)
+#endif
+    for(int64_t c = 0; c < totncells; c++){
+        for(int t = 1; t < max_nthreads; t++){
+            cell_occupation[0][c] += cell_occupation[t][c];
+        }
+    }
```

Comment:
Isn't this also a parallel prefix sum?

Reply:
Almost, it's a reduction into the "thread 0" element for each cell.

Reply:
Ohh right, of course! Instead of writing into the memory address directly, how about this (hopefully) equivalent formulation

```c
    for(int64_t c = 0; c < totncells; c++) {
         int64_t tsum = 0;
 #if defined(_OPENMP) 
#pragma omp simd for reduction(+:tsum)
#endif
        for(int t = 0; t < max_nthreads; t++){
            tsum += cell_occupation[t][c];
        }
#if  defined(_OPENMP)
#pragma omp single 
        cell_occupation[0][c] = tsum;
#endif  
      
```

Reply:
I don't think much of the runtime is going here, so I'm inclined to leave the simpler version.  Especially since we hope all this will be nuked by the sorting version!